### PR TITLE
Optimized memory usage

### PR DESCRIPTION
**Pull request by Ilya-Krylov**
* Enabled `efficient_test` that allows saving predictions to tempdir to not use CPU memory during evaluation.
* Changed `infer` method so that it does not accumulate predictions to decrease CPU memory usage
* Initialized `Image` instances by `file_path` but not numpy data

### CODE REVIEW

**Comment by Ilya-Krylov** [comment]
@evgeny-izutov @druzhkov-paul please review carefully

File: mmseg/apis/ote/apis/segmentation/inference_task.py [hunk 1]
```diff
@@ -50,6 +51,7 @@
                                                            set_hyperparams)
 from mmseg.apis.ote.apis.segmentation.configuration import OTESegmentationConfig
 from mmseg.apis.ote.apis.segmentation.ote_utils import InferenceProgressCallback
+from mmseg.core.evaluation.metrics import f_score
```
Comment (evgeny-izutov) [inline mmseg/apis/ote/apis/segmentation/inference_task.py]: Do we have a plan to replace mDice to f_score?

File: mmseg/apis/ote/apis/segmentation/inference_task.py [hunk 2]
```diff
@@ -186,67 +188,58 @@ def hook(module, input, output):
...
         pre_hook_handle.remove()
         hook_handle.remove()
 
         return dataset
-
-    @staticmethod
-    def _infer_segmentor(model: torch.nn.Module, config: Config, dataset: DatasetEntity,
+    
+    def _add_predictions_to_dataset_item(self, prediction, feature_vector, dataset_item):
+        label_dictionary = {
```
Comment (evgeny-izutov) [inline mmseg/apis/ote/apis/segmentation/inference_task.py]: Seems, it can be moved to `__init__`.

File: mmseg/apis/ote/apis/segmentation/inference_task.py [hunk 3]
```diff
@@ -186,67 +188,58 @@ def hook(module, input, output):
...
+        annotations = create_annotation_from_segmentation_map(
+            hard_prediction=hard_prediction,
+            soft_prediction=soft_prediction,
+            label_map=label_dictionary,
+        )
+        dataset_item.append_annotations(annotations=annotations)
+        if feature_vector is not None:
+            active_score = TensorEntity(name="representation_vector", numpy=feature_vector)
+            dataset_item.append_metadata_item(active_score, model=self._task_environment.model)
+            for label_index, label in label_dictionary.items():
```
Comment (evgeny-izutov) [inline mmseg/apis/ote/apis/segmentation/inference_task.py]: Seems it's mistake - it should be enabled if `dump_features==True`.

File: mmseg/apis/ote/apis/segmentation/inference_task.py (hunk 1 above)
Comment (Ilya-Krylov) [reply mmseg/apis/ote/apis/segmentation/inference_task.py]: it depends on you

File: mmseg/apis/ote/apis/segmentation/inference_task.py (hunk 3 above)
Comment (Ilya-Krylov) [reply mmseg/apis/ote/apis/segmentation/inference_task.py]: does `feature is not None` not guarantee that `dump_features==True`?

Comment (evgeny-izutov) [reply mmseg/apis/ote/apis/segmentation/inference_task.py]: Actually it depends on https://github.com/openvinotoolkit/mmsegmentation/blob/50bf757cfcf7ff07e765303d25b74794cb24cc66/mmseg/apis/ote/apis/segmentation/inference_task.py#L172-L178

Comment (Ilya-Krylov) [reply mmseg/apis/ote/apis/segmentation/inference_task.py]: Actually it does not, `dump_features = not is_evaluation` is not used.

But anyway, does `feature is not None` not guarantee that `dump_features==True`?

Comment (evgeny-izutov) [reply mmseg/apis/ote/apis/segmentation/inference_task.py]: `dump_features = not is_evaluation` is not used because the related code has been removed incompletely in this PR)

Nevertheless the answer is "no it does not depend" - it's different things. Unfortunately, they are controlled by the local variable and the method parameter with the same name `dump_features` but with different condition for the activation.

Comment (evgeny-izutov) [reply mmseg/apis/ote/apis/segmentation/inference_task.py]: The `dump_features=True` parameter of the `_infer_segmentor` method is always `True`. But  local variable `dump_features` is depend on `is_evaluation`.

Comment (Ilya-Krylov) [reply mmseg/apis/ote/apis/segmentation/inference_task.py]: I see, thanks

File: mmseg/apis/ote/apis/segmentation/inference_task.py (hunk 2 above)
Comment (Ilya-Krylov) [reply mmseg/apis/ote/apis/segmentation/inference_task.py]: fixed

File: mmseg/apis/ote/apis/segmentation/inference_task.py (hunk 3 above)
Comment (Ilya-Krylov) [reply mmseg/apis/ote/apis/segmentation/inference_task.py]: fixed

**Comment by Ilya-Krylov** [comment]
@druzhkov-paul  @evgeny-izutov please have a look

**Review by evgeny-izutov — APPROVED** [review]
So, locally it still works)

**Comment by Ilya-Krylov** [comment]
@evgeny-izutov  still ?)

**Comment by evgeny-izutov** [comment]
@Ilya-Krylov -- After achieving the state "It works" each next PR breaks the rule "If it works don't touch it" 😄

File: mmseg/apis/ote/apis/segmentation/inference_task.py [hunk 4]
```diff
@@ -265,38 +261,54 @@ def _infer_segmentor(model: torch.nn.Module, config: Config, dataset: DatasetEnt
...
         with eval_model.module.backbone.register_forward_hook(hook):
-            for data in mm_val_dataloader:
+            for i, (data, dataset_item) in enumerate(zip(mm_val_dataloader, dataset)):
                 with torch.no_grad():
                     result = eval_model(return_loss=False, output_logits=output_logits, **data)
-                eval_predictions.extend(result)
+                assert len(result) == 1
+                if eval:
+                    for k, v in mm_val_dataset.compute_stat_per_frame(result[0], frame_id=i).items():
+                        stats[k] += v
```
Comment (druzhkov-paul) [inline mmseg/apis/ote/apis/segmentation/inference_task.py]: What is it for? To compute internal mmseg metric? Do we need this?

File: mmseg/apis/ote/apis/segmentation/inference_task.py [hunk 5]
```diff
@@ -186,67 +190,59 @@ def hook(module, input, output):
...
+            soft_prediction=soft_prediction,
+            label_map=self._label_dictionary,
+        )
+        dataset_item.append_annotations(annotations=annotations)
+
+        if feature_vector is not None:
+            active_score = TensorEntity(name="representation_vector", numpy=feature_vector)
+            dataset_item.append_metadata_item(active_score, model=self._task_environment.model)
+
+        if not is_evaluation:
```
Comment (druzhkov-paul) [inline mmseg/apis/ote/apis/segmentation/inference_task.py]: I'd consider renaming `is_evalutation` to something like `save_mask_visualization`.

File: mmseg/apis/ote/apis/segmentation/inference_task.py [hunk 6]
```diff
@@ -186,67 +190,59 @@ def hook(module, input, output):
...
+                                                numpy=result_media_numpy)
+                dataset_item.append_metadata_item(result_media, model=self._task_environment.model)
+
+    def _infer_segmentor(self,
+                         model: torch.nn.Module, config: Config, dataset: DatasetEntity,
                          eval: Optional[bool] = False, metric_name: Optional[str] = 'mDice',
-                         output_logits: bool = False, dump_features: bool = True) -> Tuple[List, float]:
+                         output_logits: bool = False, dump_features: bool = True,
+                         add_predictions_to_dataset: bool = False,
+                         is_evaluation: bool = False) -> Optional[float]:
```
Comment (druzhkov-paul) [inline mmseg/apis/ote/apis/segmentation/inference_task.py]: ditto

File: mmseg/apis/ote/extension/datasets/mmdataset.py [hunk 7]
```diff
@@ -320,9 +348,7 @@ def load_dataset_items(ann_file_path: str,
         check_labels(labels_list, annot_labels)
 
     test_mode = subset in {Subset.VALIDATION, Subset.TESTING}
-    pipeline = [dict(type='LoadImageFromFile'),
-                dict(type='Normalize', mean=[0.0, 0.0, 0.0], std=[1.0, 1.0, 1.0], to_rgb=True),
```
Comment (druzhkov-paul) [inline mmseg/apis/ote/extension/datasets/mmdataset.py:324]: What about rgb<->bgr conversion?

Comment (Ilya-Krylov) [reply mmseg/apis/ote/extension/datasets/mmdataset.py:324]: https://github.com/openvinotoolkit/training_extensions/blob/master/ote_sdk/ote_sdk/entities/image.py#L81

Comment (druzhkov-paul) [reply mmseg/apis/ote/extension/datasets/mmdataset.py:324]: Please elaborate. Image is always read to RGB on the SDK side. mmseg expects BGR, where is this conversion?

Comment (Ilya-Krylov) [reply mmseg/apis/ote/extension/datasets/mmdataset.py:324]: how it was before: LoadImageFromFile reads BGR, Normalize converts it to RGB, mmseg gets RGB
now: SDK reads RGB, no conversion, mmseg gets RGB

Comment (evgeny-izutov) [reply mmseg/apis/ote/extension/datasets/mmdataset.py:324]: I thinks the key is that the mentioned `pipeline` is used for loading the annotation only. For image the path to a file is stored only here. The real image loading will be performed in another pipeline.

Comment (Ilya-Krylov) [reply mmseg/apis/ote/extension/datasets/mmdataset.py:324]: @evgeny-izutov whose opinion do you second?

Comment (evgeny-izutov) [reply mmseg/apis/ote/extension/datasets/mmdataset.py:324]: @Ilya-Krylov -- your opinion. Moreover I've tested it locally with the existing checkpoints and it achieves the same accuracy as before.

File: mmseg/apis/ote/apis/segmentation/inference_task.py (hunk 4 above)
Comment (Ilya-Krylov) [reply mmseg/apis/ote/apis/segmentation/inference_task.py]: I just tried to make changes without changing behavior.

It is set here https://github.com/openvinotoolkit/mmsegmentation/blob/56366df858d6f84641c0b3595ec03efe3bfda446/mmseg/apis/ote/apis/segmentation/train_task.py#L110 as model performance. Moreover mmseg's mDice differs from Dice score from SDK

Looks like a bug, right @druzhkov-paul ?

@evgeny-izutov ?

Comment (druzhkov-paul) [reply mmseg/apis/ote/apis/segmentation/inference_task.py]: @Ilya-Krylov I don't second your opinion on this.

I guess, we can set `score` to zero or anything dummy and leave internal metric only for best checkpoint selection.

Comment (Ilya-Krylov) [reply mmseg/apis/ote/apis/segmentation/inference_task.py]: @evgeny-izutov can we remove these computations? what do you think?

Comment (evgeny-izutov) [reply mmseg/apis/ote/apis/segmentation/inference_task.py]: @Ilya-Krylov -- I think it's mostly the question on how the [line](https://github.com/openvinotoolkit/mmsegmentation/blob/56366df858d6f84641c0b3595ec03efe3bfda446/mmseg/apis/ote/apis/segmentation/train_task.py#L115) is used later. For me it's good decision because it removes the ambiguity of metric computations inside/outside mmseg.

**Comment by Ilya-Krylov** [comment]
BMM 23487 🕐

File: mmseg/apis/ote/apis/segmentation/inference_task.py [hunk 8]
```diff
@@ -265,38 +258,29 @@ def _infer_segmentor(model: torch.nn.Module, config: Config, dataset: DatasetEnt
...
         def dummy_dump_features_hook(mod, inp, out):
-            feature_vectors.append(None)
+            pass
 
         hook = dump_features_hook if dump_features else dummy_dump_features_hook
 
         # Use a single gpu for testing. Set in both mm_val_dataloader and eval_model
         with eval_model.module.backbone.register_forward_hook(hook):
-            for data in mm_val_dataloader:
+            for i, (data, dataset_item) in enumerate(zip(mm_val_dataloader, dataset)):
```
Comment (evgeny-izutov) [inline mmseg/apis/ote/apis/segmentation/inference_task.py]: Seems the `enumerate` is never used.

File: mmseg/apis/ote/apis/segmentation/inference_task.py (hunk 5 above)
Comment (Ilya-Krylov) [reply mmseg/apis/ote/apis/segmentation/inference_task.py]: fixed

File: mmseg/apis/ote/apis/segmentation/inference_task.py (hunk 6 above)
Comment (Ilya-Krylov) [reply mmseg/apis/ote/apis/segmentation/inference_task.py]: fixed

File: mmseg/apis/ote/apis/segmentation/inference_task.py (hunk 4 above)
Comment (Ilya-Krylov) [reply mmseg/apis/ote/apis/segmentation/inference_task.py]: removed, set score to 0 and metric name to "None"

File: mmseg/apis/ote/apis/segmentation/inference_task.py [hunk 9]
```diff
@@ -74,6 +74,9 @@ def __init__(self, task_environment: TaskEnvironment):
 
         self._model_name = task_environment.model_template.name
         self._labels = task_environment.get_labels(include_empty=False)
+        self._label_dictionary = {
+            i + 1: self._labels[i] for i in range(len(self._labels))
+        }
```
Comment (druzhkov-paul) [inline mmseg/apis/ote/apis/segmentation/inference_task.py]:
```suggestion
        self._label_dictionary = dict(enumerate(self._labels, 1))
```

File: mmseg/apis/ote/apis/segmentation/inference_task.py [hunk 10]
```diff
@@ -186,67 +188,58 @@ def hook(module, input, output):
...
-
-                    result_media = ResultMediaEntity(name=f'{label.name}',
-                                                     type='Soft Prediction',
-                                                     label=label,
-                                                     annotation_scene=dataset_item.annotation_scene,
-                                                     roi=dataset_item.roi,
-                                                     numpy=result_media_numpy)
-                    dataset_item.append_metadata_item(result_media, model=self._task_environment.model)
+        self._infer_segmentor(self._model, self._config, dataset,
+                              output_logits=True, dump_features=True, add_predictions_to_dataset=True,
```
Comment (druzhkov-paul) [inline mmseg/apis/ote/apis/segmentation/inference_task.py]: In which case predictions are not saved to the dataset or features are not dumped or logits are not needed?

File: mmseg/apis/ote/apis/segmentation/inference_task.py (hunk 8 above)
Comment (Ilya-Krylov) [reply mmseg/apis/ote/apis/segmentation/inference_task.py]: fixed

File: mmseg/apis/ote/apis/segmentation/inference_task.py (hunk 10 above)
Comment (Ilya-Krylov) [reply mmseg/apis/ote/apis/segmentation/inference_task.py]: removed extra parameters

**Review by druzhkov-paul — APPROVED** [review]
